### PR TITLE
[GSoC] Add a lint rule to use string resources for preferences keys + extract deck options/filtered deck options keys

### LINKED ISSUE #10423: Uses constant in xml's key
To be sure that we avoid typos (that would be hard to find during code review), ideally, xml keys should be a string constant and not directly a string literal. Currently it's only done in preferences_reviewing.xml with

```xml
     <ListPreference
            android:defaultValue="bottom"
            android:entries="@array/answer_buttons_position_labels"
            android:entryValues="@array/answer_buttons_position"
            android:key="@string/answer_buttons_position_preference"
            android:title="@string/answer_buttons_position" />
```
which uses the value from constant.xml
```xml
<string name="answer_buttons_position_preference" translatable="false">answerButtonPosition</string>
```

This allow in AbstractFlashCardViewer.java to uses `getString(R.string.answer_buttons_position_preference)`, hence using a constant instead of a literal.

I would love to get this same pattern everywhere.

Ideally, for ease of review, a commit should change a single literal to a constant. A PR can contain multiple commit or a single one.

You can search for keys that need to be changed with the regexp `android:key="[^@]`

### PR DESCRIPTION
**Pull request by BrayanDSO**
## Fixes
Fixes #10423

## Approach
For the key extractions, just used a regex to replace the strings for keys, searching for `android:(?:key|dependency)="([^@].*?)"`. Deck options keys were replaced by `deck_options_KEYNAME_key` and filtered deck options were replaced by `filtered_deck_KEYNAME_key`. This way the change was easy to do and probably will be easy to review.

## How Has This Been Tested?

Ran lint and the unit tests created

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/res/values/preferences.xml [hunk 1]
```diff
@@ -139,4 +139,42 @@
     <string name="dev_options_enabled_msg">Enabled developer options</string>
     <!-- About -->
     <string name="about_screen_key">aboutScreen</string>
+
+    <!-- deck_options.xml -->
+    <string name="deck_options_deckConf_key">deckConf</string>
```
Comment (david-allison) [inline AnkiDroid/src/main/res/values/preferences.xml:144]: Could we move these to a separate file

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/HardcodedPreferenceKey.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import org.w3c.dom.Element
+
+/**
+ * A Lint check to prevent using hardcoded strings on preferences keys
+ */
+class HardcodedPreferenceKey : ResourceXmlDetector() {
+
+    companion object {
+        @VisibleForTesting
+        val ID = HardcodedPreferenceKey::class.simpleName!!
```
Comment (david-allison) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/HardcodedPreferenceKey.kt:32]: Use a constant, we don't want this changing if the class is renamed

File: lint-rules/src/main/java/com/ichi2/anki/lint/rules/HardcodedPreferenceKey.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    override fun visitElement(context: XmlContext, element: Element) {
+        reportAttributeIfHardcoded(context, element, "android:key")
+        reportAttributeIfHardcoded(context, element, "android:dependency")
+    }
+
+    private fun reportAttributeIfHardcoded(context: XmlContext, element: Element, attributeName: String) {
+        val attrNode = element.getAttributeNode(attributeName) ?: return
+
+        if (isHardcodedString(attrNode.value)) {
+            context.report(ISSUE, element, context.getLocation(attrNode), DESCRIPTION)
```
Comment (david-allison) [inline lint-rules/src/main/java/com/ichi2/anki/lint/rules/HardcodedPreferenceKey.kt:63]: Love seeing the location here

**Review by david-allison — CHANGES_REQUESTED** [review]
Looks great! Minor nitpicks

**Review by lukstbit — APPROVED** [review]
Everything looks ok, but please add your copyright to the two new xml files.

**Review by david-allison — APPROVED** [review]

**Pull request merged**